apps/health/lib.js

```diff
@@ -37,7 +37,36 @@ exports.readAllRecords = function(d, cb) {
     }
     idx += DB_RECORD_LEN; // +1 because we have an extra record with totals for the end of the day
   }
-}
+};
+
+// Read the entire database. There is no guarantee that the months are read in order.
+exports.readFullDatabase = function(cb) {
+  require("Storage").list(/health-[0-9]+-[0-9]+.raw/).forEach(val => {
+    console.log(val);
+    var parts = val.split('-');
+    var y = parseInt(parts[1]);
+    var mo = parseInt(parts[2].replace('.raw', ''));
+
+    exports.readAllRecords(new Date(y, mo, 1), (r) => {
+      r.date = new Date(y, mo, r.day, r.hr, r.min);
+      cb(r);
+    });
+  });
+};
+
+// Read all records per day, until the current time.
+// There may be some records for the day of the timestamp previous to the timestamp
+exports.readAllRecordsSince = function(d, cb) {
+  var currentDate = new Date().getTime();
+  var di = d;
+  while (di.getTime() < currentDate) {
+    exports.readDay(di, (r) => {
+      r.date = new Date(di.getFullYear(), di.getMonth(), di.getDate(), r.hr, r.min);
```

Comment:
@gfwilliams the samples are reaching Gadgetbridge with a 10 minute offset:

- A real-time sample had timestamp of `1691245201` in the Gadgetbridge database (14:20:01 UTC)
- The same sample when sent by this function has a timestamp of `1691244600` (14:10:00 UTC)

It looks like the Bangle database persists the samples 10 minutes "before" the activity, so the realtime samples are now offset in the Gadgetbridge database by 10 minutes.

Should we just send them to Gadgetbridge with a -10 minute offset here? Searching for a potential duplicate 10 minutes before the current sample may lead to data inconsistencies.

Reply:
Wow, good spot! Yes, it's tricky - I'd be tempted to offset in the Android lib to try and keep the API the same - having different functions returning +/- 10 min feels like it would be asking for trouble. Also looking at the code it's probably easier once you've got the Date object created anyway as it's just `d.getTime()-60000`

I guess one question is what does Gadgetbridge expect the timestamp to be? At the beginning or end of a time slot? I guess it's possible that we should actually have the live activity sample as -10 minutes? Or maybe even meet in the middle (although that doesn't seem like a great idea)

Reply:
Checking Huami devices (they have per-minute samples), it looks like they actually send the value at the end of the slot (ie. at 19:47:30, if we sync the last sample will we get will have timestamp 19:47:00).

Which I guess makes sense (at time X we had Y steps).

Changed the logic to perform the conversion here.